File: packages/python/src/kweaver/resources/context_loader.py
"""SDK resource: context-loader (MCP JSON-RPC 2.0 client).

Implements the same protocol as the TypeScript kweaverc context-loader,
providing Layer 1 (search), Layer 2 (query instances), and Layer 3
(logic properties, action info) operations against the KWeaver MCP server.
"""

from __future__ import annotations

import json
import logging
import threading
import time
from typing import TYPE_CHECKING, Any, Literal

import httpx

from kweaver._auth import _env_tls_insecure

if TYPE_CHECKING:
    from kweaver._http import HttpClient

logger = logging.getLogger("kweaver.context_loader")

MCP_PROTOCOL_VERSION = "2024-11-05"
MCP_PATH = "/api/agent-retrieval/v1/mcp"

_SESSION_TTL = 300  # seconds — sessions expire after 5 minutes

_cache_lock = threading.Lock()
_session_cache: dict[str, tuple[str, float]] = {}  # key -> (session_id, created_at)
_request_id: int = 0


def _next_id() -> int:
    global _request_id
    with _cache_lock:
        _request_id += 1
        return _request_id


def _build_mcp_url(base_url: str) -> str:
    return base_url.rstrip("/") + MCP_PATH


class ContextLoaderResource:
    """MCP context-loader client over HTTP (JSON-RPC 2.0).

    Equivalent to TypeScript src/api/context-loader.ts.
    """

    def __init__(
        self,
        base_url: str,
        access_token: str,
        kn_id: str,
        *,
        tls_insecure: bool = False,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._mcp_url = _build_mcp_url(base_url)
        self._access_token = access_token
        self._kn_id = kn_id
        self._cache_key = f"{self._mcp_url}:{kn_id}"
        self._client = httpx.Client(
            follow_redirects=True,
            verify=not (tls_insecure or _env_tls_insecure()),
        )

    def close(self) -> None:
        """Close the underlying HTTP client."""
        self._client.close()

    def __enter__(self) -> ContextLoaderResource:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

    def _build_headers(self, session_id: str | None = None) -> dict[str, str]:
        headers: dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json, text/event-stream",
            "Authorization": f"Bearer {self._access_token}",
            "X-Kn-ID": self._kn_id,
            "MCP-Protocol-Version": MCP_PROTOCOL_VERSION,
        }
        if session_id:
            headers["MCP-Session-Id"] = session_id
        return headers

    def _ensure_session(self) -> str:
        with _cache_lock:
            cached = _session_cache.get(self._cache_key)
            if cached:
                session_id, created_at = cached
                if time.time() - created_at < _SESSION_TTL:
                    return session_id
                # Expired — remove stale entry
                _session_cache.pop(self._cache_key, None)

        init_body = json.dumps({
            "jsonrpc": "2.0",
            "id": _next_id(),
            "method": "initialize",
            "params": {
                "protocolVersion": MCP_PROTOCOL_VERSION,
                "capabilities": {},
                "clientInfo": {"name": "kweaver-sdk", "version": "0.1.0"},
            },
        })

        resp = self._client.post(
            self._mcp_url,
            content=init_body,
            headers=self._build_headers(),
            timeout=30.0,
        )
        if resp.status_code >= 400:
            raise RuntimeError(f"MCP initialize failed with HTTP {resp.status_code}")

        session_id = (
            resp.headers.get("MCP-Session-Id")
            or resp.headers.get("mcp-session-id")
        )
        if not session_id:
            raise RuntimeError(
                "MCP server did not return MCP-Session-Id. "
                "The server may require session initialization."
            )

        notif_body = json.dumps({
            "jsonrpc": "2.0",
            "method": "notifications/initialized",
        })
        self._client.post(
            self._mcp_url,
            content=notif_body,
            headers=self._build_headers(session_id),
            timeout=10.0,
        )

        with _cache_lock:
            _session_cache[self._cache_key] = (session_id, time.time())
        return session_id

    def _call_method(self, method: str, params: dict[str, Any] | None = None) -> Any:
        session_id = self._ensure_session()
        body: dict[str, Any] = {
            "jsonrpc": "2.0",
            "method": method,
            "id": _next_id(),
        }
        if params:
            body["params"] = params

        resp = self._client.post(
            self._mcp_url,
            content=json.dumps(body),
            headers=self._build_headers(session_id),
            timeout=60.0,
        )
        if resp.status_code >= 400:
            raise RuntimeError(f"MCP {method} failed with HTTP {resp.status_code}")

        parsed = resp.json()
        if error := parsed.get("error"):
            raise RuntimeError(f"Context-loader error: {error.get('message', error)}")

        if "result" in parsed:
            return parsed["result"]

        raise RuntimeError(f"Context-loader returned no result for method {method}")

    def _call_tool(self, tool_name: str, args: dict[str, Any]) -> Any:
        session_id = self._ensure_session()
        body = {
            "jsonrpc": "2.0",
            "method": "tools/call",
            "params": {"name": tool_name, "arguments": args},
            "id": _next_id(),
        }

        resp = self._client.post(
            self._mcp_url,
            content=json.dumps(body),
            headers=self._build_headers(session_id),
            timeout=60.0,
        )
        if resp.status_code >= 400:
            raise RuntimeError(f"MCP tools/call '{tool_name}' failed with HTTP {resp.status_code}")

        parsed = resp.json()

        if _is_missing_input_params(parsed):
            raise RuntimeError(_format_missing_input_params(parsed))

        if error := parsed.get("error"):
            data = error.get("data")
            if _is_missing_input_params(data):
                raise RuntimeError(_format_missing_input_params(data))
            raise RuntimeError(f"Context-loader error: {error.get('message', error)}")

        if "result" in parsed:
            result = parsed["result"]
            content = result.get("content") if isinstance(result, dict) else None
            if isinstance(content, list) and content:
                first = content[0]
                text = first.get("text") if isinstance(first, dict) else None
                if isinstance(text, str):
                    try:
                        return json.loads(text)
                    except json.JSONDecodeError:
                        return {"raw": text}
            return result

        if any(k in parsed for k in ("object_types", "concepts", "datas", "entries", "_dynamic_tools")):
            return parsed

        raise RuntimeError(f"Context-loader returned no result for tool {tool_name}")

    # ── Layer 1 ─────────────────────────────────────────────────────────────

    def call_tool(self, tool_name: str, args: dict[str, Any]) -> Any:
        """Call an arbitrary MCP tool by name."""
        return self._call_tool(tool_name, args)

    def search_schema(
        self,
        query: str,
        *,
        response_format: Literal["json", "toon"] = "json",
        search_scope: dict[str, bool] | None = None,
        max_concepts: int | None = None,
        schema_brief: bool | None = None,
        enable_rerank: bool | None = None,
    ) -> dict[str, Any]:
        """Search schema via the MCP search_schema tool."""
        args: dict[str, Any] = {
            "query": query,
            "response_format": response_format,
        }
        if search_scope is not None:
            args["search_scope"] = search_scope
        if max_concepts is not None:
            args["max_concepts"] = max_concepts
        if schema_brief is not None:
            args["schema_brief"] = schema_brief
        if enable_rerank is not None:
            args["enable_rerank"] = enable_rerank
        return self._call_tool("search_schema", args)

    # ── Layer 2 ─────────────────────────────────────────────────────────────

    def query_object_instance(
        self, ot_id: str, condition: dict[str, Any], *, limit: int = 20
    ) -> dict[str, Any]:
        """Query object instances — returns datas with _instance_identity."""
        return self._call_tool("query_object_instance", {
            "ot_id": ot_id,
            "condition": condition,
            "limit": limit,
        })

    def query_instance_subgraph(
        self, relation_type_paths: list[dict[str, Any]]
    ) -> dict[str, Any]:
        """Query instance subgraph — returns entries with nested _instance_identity."""
        return self._call_tool("query_instance_subgraph", {
            "relation_type_paths": relation_type_paths,
        })

    # ── Layer 3 ─────────────────────────────────────────────────────────────

    def get_logic_properties_values(
        self,
        ot_id: str,
        query: str,
        instance_identities: list[dict[str, Any]],
        properties: list[str],
        additional_context: str | None = None,
    ) -> dict[str, Any]:
        """Get logic property values. Raises on MISSING_INPUT_PARAMS."""
        args: dict[str, Any] = {
            "ot_id": ot_id,
            "query": query,
            "_instance_identities": instance_identities,
            "properties": properties,
        }
        if additional_context is not None:
            args["additional_context"] = additional_context
        return self._call_tool("get_logic_properties_values", args)

    def get_action_info(
        self, at_id: str, instance_identity: dict[str, Any]
    ) -> dict[str, Any]:
        """Get action info — returns _dynamic_tools."""
        return self._call_tool("get_action_info", {
            "at_id": at_id,
            "_instance_identity": instance_identity,
        })

    def find_skills(
        self,
        object_type_id: str,
        *,
        skill_query: str | None = None,
        top_k: int | None = None,
        instance_identities: list[dict[str, Any]] | None = None,
        response_format: Literal["json", "toon"] | None = None,
    ) -> dict[str, Any]:
        """Recall skills attached to an object type via the find_skills MCP tool.

        Args:
            object_type_id: Required object type id whose skills should be recalled.
            skill_query: Optional natural-language query to narrow recall.
            top_k: Optional 1..20 cap on returned skills.
            instance_identities: Optional list of instance identities to scope recall.
            response_format: Optional output format ("json" or "toon").

        Returns:
            Tool payload with ``entries`` and an optional ``message``.
        """
        if not object_type_id:
            raise ValueError("find_skills: object_type_id is required.")
        if top_k is not None and not (1 <= top_k <= 20):
            raise ValueError("find_skills: top_k must be between 1 and 20.")
        args: dict[str, Any] = {"object_type_id": object_type_id}
        if response_format is not None:
            args["response_format"] = response_format
        if instance_identities is not None:
            args["instance_identities"] = instance_identities
        if skill_query is not None:
            args["skill_query"] = skill_query
        if top_k is not None:
            args["top_k"] = top_k
        return self._call_tool("find_skills", args)

    # ── MCP introspection ────────────────────────────────────────────────────

    def list_tools(self, cursor: str | None = None) -> dict[str, Any]:
        params = {"cursor": cursor} if cursor else {}
        return self._call_method("tools/list", params or None)

    def list_resources(self, cursor: str | None = None) -> dict[str, Any]:
        params = {"cursor": cursor} if cursor else {}
        return self._call_method("resources/list", params or None)

    def read_resource(self, uri: str) -> dict[str, Any]:
        return self._call_method("resources/read", {"uri": uri})

    def list_resource_templates(self, cursor: str | None = None) -> dict[str, Any]:
        params = {"cursor": cursor} if cursor else {}
        return self._call_method("resources/templates/list", params or None)

    def list_prompts(self, cursor: str | None = None) -> dict[str, Any]:
        params = {"cursor": cursor} if cursor else {}
        return self._call_method("prompts/list", params or None)

    def get_prompt(self, name: str, args: dict[str, Any] | None = None) -> dict[str, Any]:
        params: dict[str, Any] = {"name": name}
        if args:
            params["arguments"] = args
        return self._call_method("prompts/get", params)

    @classmethod
    def clear_session_cache(cls) -> None:
        """Clear the session cache (useful for testing)."""
        _session_cache.clear()


def _is_missing_input_params(obj: Any) -> bool:
    return (
        isinstance(obj, dict)
        and obj.get("error_code") == "MISSING_INPUT_PARAMS"
    )


def _format_missing_input_params(err: dict[str, Any]) -> str:
    lines = [
        f"MISSING_INPUT_PARAMS: {err.get('message', '')}",
        "Add the following to additional_context and retry:",
    ]
    for m in err.get("missing", []):
        for p in m.get("params", []):
            hint = p.get("hint")
            if hint:
                lines.append(f"  - {p['name']}: {hint}")
    return "\n".join(lines)

File: packages/typescript/src/api/context-loader.ts
import { isNoAuth } from "../config/no-auth.js";
import { fetchTextOrThrow } from "../utils/http.js";

/** Per-call options for context-loader MCP. */
export interface ContextLoaderCallOptions {
  mcpUrl: string;
  knId: string;
  accessToken: string;
}

const MCP_PROTOCOL_VERSION = "2024-11-05";

const SESSION_TTL_MS = 300_000; // 5 minutes
const sessionCache = new Map<string, { id: string; createdAt: number }>();

function sessionKey(options: ContextLoaderCallOptions): string {
  return `${options.mcpUrl}:${options.knId}`;
}

function buildHeaders(options: ContextLoaderCallOptions, sessionId?: string): Record<string, string> {
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "application/json, text/event-stream",
    "X-Kn-ID": options.knId,
    "MCP-Protocol-Version": MCP_PROTOCOL_VERSION,
  };
  if (!isNoAuth(options.accessToken)) {
    headers.Authorization = `Bearer ${options.accessToken}`;
  }
  if (sessionId) {
    headers["MCP-Session-Id"] = sessionId;
  }
  return headers;
}

async function ensureSession(options: ContextLoaderCallOptions): Promise<string> {
  const key = sessionKey(options);
  const cached = sessionCache.get(key);
  if (cached && Date.now() - cached.createdAt < SESSION_TTL_MS) return cached.id;
  // Remove stale entry if expired
  if (cached) sessionCache.delete(key);

  const initBody = JSON.stringify({
    jsonrpc: "2.0",
    id: 1,
    method: "initialize",
    params: {
      protocolVersion: MCP_PROTOCOL_VERSION,
      capabilities: {},
      clientInfo: { name: "kweaver-caller", version: "0.1.0" },
    },
  });

  const { response, body } = await fetchTextOrThrow(options.mcpUrl, {
    method: "POST",
    headers: buildHeaders(options),
    body: initBody,
  });

  const sessionId = response.headers.get("MCP-Session-Id") ?? response.headers.get("mcp-session-id");
  if (!sessionId) {
    throw new Error(
      "MCP server did not return MCP-Session-Id. The server may require session initialization."
    );
  }

  const initNotifBody = JSON.stringify({
    jsonrpc: "2.0",
    method: "notifications/initialized",
  });

  await fetchTextOrThrow(options.mcpUrl, {
    method: "POST",
    headers: buildHeaders(options, sessionId),
    body: initNotifBody,
  });

  sessionCache.set(key, { id: sessionId, createdAt: Date.now() });
  return sessionId;
}

export interface SearchSchemaScope {
  include_object_types?: boolean;
  include_relation_types?: boolean;
  include_action_types?: boolean;
  include_metric_types?: boolean;
}

/** Layer 1: search_schema arguments. */
export interface SearchSchemaArgs {
  query: string;
  response_format?: "json" | "toon";
  search_scope?: SearchSchemaScope;
  max_concepts?: number;
  schema_brief?: boolean;
  enable_rerank?: boolean;
}

/** Layer 1: search_schema result. */
export interface SearchSchemaResult {
  object_types?: unknown[];
  relation_types?: unknown[];
  action_types?: unknown[];
  metric_types?: unknown[];
  raw?: string;
}

/** Condition for query_object_instance and query_instance_subgraph. */
export interface ConditionSpec {
  operation: "and" | "or";
  sub_conditions: Array<{
    field?: string;
    operation?: string;
    value_from?: "const";
    value?: unknown;
  }>;
}

/** Layer 2: query_object_instance arguments. */
export interface QueryObjectInstanceArgs {
  ot_id: string;
  limit?: number;
  condition: ConditionSpec;
}

/** Layer 2: query_instance_subgraph relation type path. */
export interface RelationTypePath {
  object_types: Array<{
    id: string;
    condition: ConditionSpec;
    limit?: number;
  }>;
  relation_types: Array<{
    relation_type_id: string;
    source_object_type_id: string;
    target_object_type_id: string;
  }>;
}

/** Layer 2: query_instance_subgraph arguments. */
export interface QueryInstanceSubgraphArgs {
  relation_type_paths: RelationTypePath[];
}

/** Layer 3: get_logic_properties_values arguments. */
export interface GetLogicPropertiesValuesArgs {
  ot_id: string;
  query: string;
  _instance_identities: Record<string, string>[];
  properties: string[];
  additional_context?: string;
}

/** Layer 3: get_action_info arguments. */
export interface GetActionInfoArgs {
  at_id: string;
  _instance_identity: Record<string, string>;
}

/** Layer 3: find_skills arguments (object_type_id is required). */
export interface FindSkillsArgs {
  object_type_id: string;
  response_format?: "json" | "toon";
  instance_identities?: Record<string, unknown>[];
  skill_query?: string;
  /** 1..20, default 10 on the server side. */
  top_k?: number;
}

/** Layer 3: find_skills result. */
export interface FindSkillsResult {
  entries: Array<{ skill_id: string; name: string; description?: string }>;
  message?: string;
}

/** Error when get_logic_properties_values returns MISSING_INPUT_PARAMS. */
export interface MissingInputParamsError {
  error_code: "MISSING_INPUT_PARAMS";
  message: string;
  missing: Array<{
    property: string;
    params: Array<{ name: string; type: string; hint: string }>;
  }>;
}

function isMissingInputParams(obj: unknown): obj is MissingInputParamsError {
  return (
    typeof obj === "object" &&
    obj !== null &&
    (obj as Record<string, unknown>).error_code === "MISSING_INPUT_PARAMS"
  );
}

/** Build retry hint from MISSING_INPUT_PARAMS response. */
export function formatMissingInputParamsHint(err: MissingInputParamsError): string {
  const lines: string[] = [
    "MISSING_INPUT_PARAMS: " + err.message,
    "Add the following to additional_context and retry:",
  ];
  for (const m of err.missing ?? []) {
    for (const p of m.params ?? []) {
      if (p.hint) lines.push(`  - ${p.name}: ${p.hint}`);
    }
  }
  return lines.join("\n");
}

/** Guardrail: value_from only supports "const"; must appear with value. */
export function validateCondition(condition: unknown): void {
  if (!condition || typeof condition !== "object") return;
  const c = condition as Record<string, unknown>;
  const sub = c.sub_conditions;
  if (Array.isArray(sub)) {
    for (const s of sub) {
      if (s && typeof s === "object") {
        const sc = s as Record<string, unknown>;
        if (sc.value_from !== undefined && sc.value_from !== "const") {
          throw new Error(
            'Condition value_from only supports "const". Must use value_from: "const" with value.'
          );
        }
        if (sc.value_from === "const" && sc.value === undefined) {
          throw new Error('When value_from is "const", value must be provided.');
        }
      }
    }
  }
  if (c.value_from !== undefined && c.value_from !== "const") {
    throw new Error('Condition value_from only supports "const".');
  }
}

/** Guardrail: _instance_identity must be a plain object (from Layer 2 output, not fabricated). */
export function validateInstanceIdentity(v: unknown, label: string): void {
  if (v === null || typeof v !== "object" || Array.isArray(v)) {
    throw new Error(
      `${label} must be a plain object from Layer 2 query result (_instance_identity). Do not fabricate.`
    );
  }
}

/** Guardrail: _instance_identities must be array of plain objects from Layer 2. */
export function validateInstanceIdentities(v: unknown): void {
  if (!Array.isArray(v)) {
    throw new Error("_instance_identities must be an array from Layer 2 query result.");
  }
  for (let i = 0; i < v.length; i += 1) {
    validateInstanceIdentity(v[i], `_instance_identities[${i}]`);
  }
}

let requestId = 0;

/** Call a generic MCP JSON-RPC method (e.g. tools/list, resources/list). Returns result as-is. */
async function callMcpMethod(
  options: ContextLoaderCallOptions,
  method: string,
  params: Record<string, unknown> = {}
): Promise<unknown> {
  const sessionId = await ensureSession(options);
  const id = (requestId += 1);

  const body = JSON.stringify({
    jsonrpc: "2.0",
    method,
    params: Object.keys(params).length > 0 ? params : undefined,
    id,
  });

  const { body: responseBody } = await fetchTextOrThrow(options.mcpUrl, {
    method: "POST",
    headers: buildHeaders(options, sessionId),
    body,
  });

  let parsed: unknown;
  try {
    parsed = JSON.parse(responseBody);
  } catch {
    throw new Error(`Context-loader returned invalid JSON: ${responseBody.slice(0, 200)}`);
  }

  const rpc = parsed as { result?: unknown; error?: { code: number; message: string } };
  if (rpc.error) {
    throw new Error(`Context-loader error: ${rpc.error.message}`);
  }

  if (rpc.result !== undefined) {
    return rpc.result;
  }

  throw new Error("Context-loader returned no result");
}

export async function callTool(
  options: ContextLoaderCallOptions,
  toolName: string,
  args: Record<string, unknown>
): Promise<unknown> {
  const sessionId = await ensureSession(options);
  const id = (requestId += 1);

  const body = JSON.stringify({
    jsonrpc: "2.0",
    method: "tools/call",
    params: { name: toolName, arguments: args as Record<string, unknown> },
    id,
  });

  const { body: responseBody } = await fetchTextOrThrow(options.mcpUrl, {
    method: "POST",
    headers: buildHeaders(options, sessionId),
    body,
  });

  let parsed: unknown;
  try {
    parsed = JSON.parse(responseBody);
  } catch {
    throw new Error(`Context-loader returned invalid JSON: ${responseBody.slice(0, 200)}`);
  }

  if (isMissingInputParams(parsed)) {
    throw new Error(formatMissingInputParamsHint(parsed));
  }

  const rpc = parsed as { result?: unknown; error?: { code: number; message: string; data?: unknown } };
  if (rpc.error) {
    const data = rpc.error.data;
    if (isMissingInputParams(data)) {
      throw new Error(formatMissingInputParamsHint(data));
    }
    throw new Error(`Context-loader error: ${rpc.error.message}`);
  }

  const result = rpc.result;
  if (result !== undefined) {
    const content = (result as Record<string, unknown>).content;
    if (Array.isArray(content) && content.length > 0) {
      const first = content[0] as Record<string, unknown>;
      const text = first?.text;
      if (typeof text === "string") {
        try {
          return JSON.parse(text) as unknown;
        } catch {
          return { raw: text };
        }
      }
    }
    return result;
  }

  const direct = parsed as Record<string, unknown>;
  if (
    "object_types" in direct ||
    "concepts" in direct ||
    "datas" in direct ||
    "entries" in direct ||
    "_dynamic_tools" in direct
  ) {
    return parsed;
  }

  throw new Error("Context-loader returned no result");
}

/** Layer 1: search_schema. Returns object_types, relation_types, action_types, metric_types. */
export async function searchSchema(
  options: ContextLoaderCallOptions,
  args: SearchSchemaArgs
): Promise<SearchSchemaResult> {
  const toolArgs: Record<string, unknown> = {
    query: args.query,
    response_format: args.response_format ?? "json",
  };
  if (args.search_scope !== undefined) toolArgs.search_scope = args.search_scope;
  if (args.max_concepts !== undefined) toolArgs.max_concepts = args.max_concepts;
  if (args.schema_brief !== undefined) toolArgs.schema_brief = args.schema_brief;
  if (args.enable_rerank !== undefined) toolArgs.enable_rerank = args.enable_rerank;
  return (await callTool(options, "search_schema", toolArgs)) as SearchSchemaResult;
}

/** Layer 2: query_object_instance. Returns datas with _instance_identity. */
export async function queryObjectInstance(
  options: ContextLoaderCallOptions,
  args: QueryObjectInstanceArgs
): Promise<unknown> {
  validateCondition(args.condition);
  return callTool(options, "query_object_instance", { ...args });
}

/** Layer 2: query_instance_subgraph. Returns entries with nested _instance_identity. */
export async function queryInstanceSubgraph(
  options: ContextLoaderCallOptions,
  args: QueryInstanceSubgraphArgs
): Promise<unknown> {
  for (const path of args.relation_type_paths) {
    for (const ot of path.object_types) {
      validateCondition(ot.condition);
    }
  }
  return callTool(options, "query_instance_subgraph", { ...args });
}

/** Layer 3: get_logic_properties_values. Throws with retry hint on MISSING_INPUT_PARAMS. */
export async function getLogicPropertiesValues(
  options: ContextLoaderCallOptions,
  args: GetLogicPropertiesValuesArgs
): Promise<unknown> {
  validateInstanceIdentities(args._instance_identities);
  return callTool(options, "get_logic_properties_values", { ...args });
}

/** Layer 3: get_action_info. Returns _dynamic_tools. */
export async function getActionInfo(
  options: ContextLoaderCallOptions,
  args: GetActionInfoArgs
): Promise<unknown> {
  validateInstanceIdentity(args._instance_identity, "_instance_identity");
  return callTool(options, "get_action_info", { ...args });
}

/**
 * Layer 3: find_skills. Recall skills attached to an object type or
 * (optionally) narrowed to specific instances.
 */
export async function findSkills(
  options: ContextLoaderCallOptions,
  args: FindSkillsArgs
): Promise<FindSkillsResult> {
  if (!args.object_type_id || typeof args.object_type_id !== "string") {
    throw new Error("find_skills: object_type_id is required.");
  }
  if (args.top_k !== undefined && (args.top_k < 1 || args.top_k > 20)) {
    throw new Error("find_skills: top_k must be between 1 and 20.");
  }
  if (args.instance_identities !== undefined) {
    validateInstanceIdentities(args.instance_identities);
  }
  const toolArgs: Record<string, unknown> = {
    object_type_id: args.object_type_id,
  };
  if (args.response_format !== undefined) toolArgs.response_format = args.response_format;
  if (args.instance_identities !== undefined) toolArgs.instance_identities = args.instance_identities;
  if (args.skill_query !== undefined) toolArgs.skill_query = args.skill_query;
  if (args.top_k !== undefined) toolArgs.top_k = args.top_k;
  return (await callTool(options, "find_skills", toolArgs)) as FindSkillsResult;
}

/** MCP tools/list. Returns list of available tools. */
export async function listTools(
  options: ContextLoaderCallOptions,
  params?: { cursor?: string }
): Promise<unknown> {
  return callMcpMethod(options, "tools/list", params ? { cursor: params.cursor } : {});
}

/** MCP resources/list. Returns list of available resources. */
export async function listResources(
  options: ContextLoaderCallOptions,
  params?: { cursor?: string }
): Promise<unknown> {
  return callMcpMethod(options, "resources/list", params ? { cursor: params.cursor } : {});
}

/** MCP resources/read. Returns resource content by URI. */
export async function readResource(
  options: ContextLoaderCallOptions,
  uri: string
): Promise<unknown> {
  return callMcpMethod(options, "resources/read", { uri });
}

/** MCP resources/templates/list. Returns list of resource templates. */
export async function listResourceTemplates(
  options: ContextLoaderCallOptions,
  params?: { cursor?: string }
): Promise<unknown> {
  return callMcpMethod(options, "resources/templates/list", params ? { cursor: params.cursor } : {});
}

/** MCP prompts/list. Returns list of available prompts. */
export async function listPrompts(
  options: ContextLoaderCallOptions,
  params?: { cursor?: string }
): Promise<unknown> {
  return callMcpMethod(options, "prompts/list", params ? { cursor: params.cursor } : {});
}

/** MCP prompts/get. Returns prompt by name with optional arguments. */
export async function getPrompt(
  options: ContextLoaderCallOptions,
  name: string,
  args?: Record<string, unknown>
): Promise<unknown> {
  return callMcpMethod(options, "prompts/get", args ? { name, arguments: args } : { name });
}
